src/components/grid.js

```diff
@@ -86,87 +122,109 @@ class Grid extends Component {
         });
     }
 
-    checkPlayerInventory(itemToCheck, required) {
-        const inventory = this.props.player.inventory[itemToCheck];
+    checkRemainingEnemies() {
+        const npcSquares = [];
+
+        this.state.gridLayout.forEach((row) => {
+            row.forEach((square) => {
+                if (square.npc) {
+                    npcSquares.push(square);
+                }
+            });
+        });
+
+        const remainingEnemies = npcSquares.filter((square) => {
+            if (!square.npc.actions[0].requirements) return;
+
+            return square.npc.actions[0].requirements[0].cleared === false;
+        });
+
+        return remainingEnemies.length;
+    }
+
+    checkPlayerInventory(itemToCheck, amountRequired) {
+        const name = this.props.player.getName();
+        const inventory = this.props.player.getInventory(itemToCheck);
 
-        if (inventory === 0) {
+        if (!inventory) {
             this.props.updateMessage(
-                'You must first bring me back 3 report cards before you are allowed to pass!'
+                `"Greetings, ${name}! You must first bring me back 3 report cards before you are allowed to pass!"`,
+                'dialogue'
             );
             return false;
-        } else if (inventory < required) {
+        } else if (inventory < amountRequired) {
             this.props.updateMessage(
-                `It looks like you're almost there! Only ${
-                    required - inventory[0].quantity
-                } more report card${
-                    inventory[0].quantity !== 1 ? '' : 's'
-                } to go!`
+                `"Oh, you're quite on your way, ${name}! You only need ${
+                    amountRequired - inventory
+                } more report card${inventory !== 1 ? '' : 's'}!"`,
+                'dialogue'
             );
             return false;
         }
 
         return true;
     }
 
-    interact(row, column) {
+    handleInteraction(row, column) {
         const gridSpace = this.state.gridLayout[row][column];
+        const { updateMessage } = this.props;
 
         if (gridSpace.npc && !gridSpace.npc.isInTheMiddleOfAction) {
             const currentAction = gridSpace.npc.interact();
+            const cancel = gridSpace.npc.cancel();
```

Comment:
It looks like maybe you're meaning to save a reference to the cancel method to call later, but since it's actually getting called here, I believe `cancel` is being saved as `undefined` because `npc.cancel()` doesn't return anything.